Fix inplace ops on Partial DTensors to preserve aliasing semantics

RohitRathore1 opened the pull request:
Fixes #163374. 

Here is the output from reproducible code:

```
W1006 09:09:26.329000 2457 /home/fedora/github/pytorch/torch/distributed/run.py:811] 
W1006 09:09:26.329000 2457 /home/fedora/github/pytorch/torch/distributed/run.py:811] *****************************************
W1006 09:09:26.329000 2457 /home/fedora/github/pytorch/torch/distributed/run.py:811] Setting OMP_NUM_THREADS environment variable for each process to be 1 in default, to avoid your system being overloaded, please further tune the variable for optimal performance in your application as needed. 
W1006 09:09:26.329000 2457 /home/fedora/github/pytorch/torch/distributed/run.py:811] *****************************************
  aten::clamp_(dt: f32[][R], None, 2)
    redistribute_input(0, [P] -> [R])
      redistribute_input(t: f32[], [P] -> [R])
        _c10d_functional::all_reduce(t: f32[], sum, 0)
        _c10d_functional::wait_tensor(t: f32[])
    aten::clamp_(t: f32[], None, 2)
    aten::view(t: f32[], [])
(Replicate(),)
tensor(2., device='cuda:0')
```

The behavior is now matching what you were expecting in issue #163374:

Expected behavior (from the issue):
  1. Placement should change from Partial(sum) to Replicate()
  2. Value should be tensor(2.) instead of tensor(144.)

  Actual output from this build:
  1. (Replicate(),) - placement is correct
  2. tensor(2., device='cuda:0') - value is correct

so the inplace operation now properly redistributes the partial DTensor to replicate before performing the clamp snd maintains the correct aliasing semantics. It also produces the expected clamped value. 

cc: @SherlockNoMad @ezyang @janeyx99 @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci

RohitRathore1 committed:
Fix inplace ops on Partial DTensors to preserve aliasing semantics

RohitRathore1 commented:
@pytorchbot label "topic: not user facing"

RohitRathore1 committed:
Address feedback

ezyang reviewed (changes requested):
I don't think this is right. The problem is if there is a view into the partial tensor; just overwriting the local tensor is not enough to cause the views to update. In fact, I tend to think you are just toast if you need a redistribute on the tensor being inplace'd. I'd look more positively on something that just errors when this happens.

ezyang commented:
sure

RohitRathore1 committed:
Error on inplace ops requiring redistribution to preserve aliasing
Per @ezyang's feedback, inplace operations that require redistribution
(e.g., Partial -> Replicate) now raise a RuntimeError instead of trying
to work around the issue by overwriting _local_tensor.

ezyang commented on torch/distributed/tensor/_dispatch.py:
```diff
@@ -288,15 +288,30 @@ def default_tensor(spec: DTensorSpec) -> torch.Tensor:
+
+                # update the spec for all inplace ops to handle placement changes
+                # that don't require redistribution
+                args[0]._spec = output_spec
```
I think it's still OK (and in fact preferred) for the squeeze_ logic to live inside the squeeze_ branch. Most inplace ops do NOT change the tensor meta. So because no redistribution could have occurred, the input cannot have a placement change.

ezyang commented:
Just a small tweak please, thanks

RohitRathore1 committed:
Address suggestions

RohitRathore1 commented:
> Just a small tweak please, thanks

Done, thanks!

ezyang commented:
need to update tests

RohitRathore1 commented:
> need to update tests

@ezyang you are referring to failing test or do I need to update existing test?

ezyang commented:
your ci failed, fix it

RohitRathore1 committed:
Fix CI: exempt squeeze_.dim from placement change check

RohitRathore1 commented:
@ezyang I've fixed the CI failure. The issue was that the placement change check was happening before the `squeeze_.dim` special case, so I moved it into the else block to exempt `squeeze_.dim`. I also removed the spec update for other inplace ops as you mentioned since most inplace ops don't change tensor meta and no redistribution occurs when placements match, the spec update isn't needed. Let me know if there's anything else that needs to be addressed.

ezyang commented:
@pytorchbot merge
